brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,171 @@
+ public:
+    TorLauncherDelegate() {}
+    ~TorLauncherDelegate() override {}
+
+    void RunAsyncSafe() override {
+#if defined(OS_POSIX)
+      for (size_t i = 0; i < 2; ++i)
+        close(pipehack[i]);
+#endif
+      // TODO(darkdh): doesn't prevent ctrl+c propagated by terminal
+      struct sigaction action;
+      memset(&action, 0, sizeof(action));
+      action.sa_handler = SIG_IGN;
+      sigaction(SIGINT, &action, NULL);
+    }
+ private:
+    DISALLOW_COPY_AND_ASSIGN(TorLauncherDelegate);
+};
+
+TorLauncherImpl::TorLauncherImpl(
+    std::unique_ptr<service_manager::ServiceContextRef> service_ref)
+    : service_ref_(std::move(service_ref)) {
+  SetupPipeHack();
+}
+
+TorLauncherImpl::~TorLauncherImpl() {
+  if (tor_process_.IsValid()) {
+#if defined(OS_POSIX)
+    for (size_t i = 0; i < 2; ++i)
+      close(pipehack[i]);
+#endif
+    tor_process_.Terminate(0, false);
+#if defined(OS_MACOSX)
+    base::PostTaskWithTraits(
+        FROM_HERE, {base::MayBlock(), base::TaskPriority::BACKGROUND},
+        base::BindOnce(&base::EnsureProcessTerminated, Passed(&tor_process_)));
+#else
+    base::EnsureProcessTerminated(std::move(tor_process_));
+#endif
+  }
+}
+
+void TorLauncherImpl::Launch(const base::FilePath& tor_bin,
+                             const std::vector<std::string>& args,
+                             LaunchCallback callback) {
+  base::CommandLine cmdline(tor_bin);
+  for (size_t i = 0; i < args.size(); ++i) {
+    cmdline.AppendArg(args[i]);
+  }
+  base::LaunchOptions launchopts;
+  TorLauncherDelegate tor_launcher_delegate;
+  launchopts.pre_exec_delegate = &tor_launcher_delegate;
+#if defined(OS_LINUX)
+  launchopts.kill_on_parent_death = true;
+#endif
+  tor_process_ = base::LaunchProcess(cmdline, launchopts);
+
+  bool result;
+  // TODO(darkdh): return success when tor connected to tor network
+  if (tor_process_.IsValid())
+    result = true;
+  else
+    result = false;
+
+  if (callback)
+    std::move(callback).Run(result);
+
+  if (!child_monitor_thread_.get()) {
+    child_monitor_thread_.reset(new base::Thread("child_monitor_thread"));
+    if (!child_monitor_thread_->Start()) {
+      NOTREACHED();
+    }
+
+    child_monitor_thread_->task_runner()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&TorLauncherImpl::MonitorChild, base::Unretained(this)));
+  }
+}
+
+void TorLauncherImpl::SetCrashHandler(SetCrashHandlerCallback callback) {
+  crash_handler_callback_ = std::move(callback);
+}
+
+void TorLauncherImpl::MonitorChild() {
+#if defined(OS_POSIX)
+  pid_t pid;
+
+  while (1) {
+      if (read(pipehack[0], &pid, sizeof(pid_t)) > 0) {
+        if (crash_handler_callback_)
+          std::move(crash_handler_callback_).Run(pid);
+        continue;
+      } else {
+        // pipes closed
+        break;
+      }
+  }
+#elif defined(OS_WINDOWS)
+  WaitForSingleObject(tor_process_.Handle(), INFINITE);
+  if (crash_handler_callback_)
+    std::move(crash_handler_callback_).Run(pid);
```

Comment:
What happens when tor exits because the user is exiting the browser and we cause the tor subprocess to terminate?  Does the crash handler callback still run in that case?

Reply:
This crash handler callback will not run on POSIX, I haven't tested on Windows yet.